File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
public:
  int maxAreaOfIsland(vector<vector<int>> &grid) {
    int n = grid.size();
    int m = grid[0].size();
    vector<vector<int>> visited(n, vector<int>(m, -1));
    int res = 0;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (visited[i][j] == -1 && grid[i][j] == 1) {
          res = max(dfs(grid, i, j, visited), res);
        }
      }
    }
    return res;
  }
  int dfs(const vector<vector<int>> &grid, int i, int j,
          vector<vector<int>> &visited) {
    if (i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size() ||
        grid[i][j] == 0)
      return 0;
    if (visited[i][j] == 1) {
      return 0;
    }
    visited[i][j] = 1;
    return 1 + dfs(grid, i + 1, j, visited) + dfs(grid, i - 1, j, visited) +
           dfs(grid, i, j - 1, visited) + dfs(grid, i, j + 1, visited);
  }

};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height)-1
        lmax, rmax = height[l], height[r]
        res = 0
        while l < r:
            if lmax < rmax:
                l += 1
                if lmax - height[l] > 0:
                    res += lmax - height[l]

                lmax = max(height[l], lmax)
            else:
                r-=1
                if rmax - height[r] > 0:
                    res += rmax- height[r]
                rmax = max(height[r],rmax)
        return res

